src/track-promise.cc

```diff
@@ -0,0 +1,67 @@
+#include "env.h"
+#include "env-inl.h"
+#include "node_internals.h"
+
+namespace node {
+
+using v8::Exception;
+using v8::Function;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::Persistent;
+using v8::Value;
+using v8::WeakCallbackData;
+
+typedef void (*FreeCallback)(Local<Object> object, Local<Function> fn);
+
+class TrackPromise {
+ public:
+  static TrackPromise* New(Isolate* isolate, Local<Object> object);
+  inline Persistent<Object>* persistent();
+ private:
+  static void WeakCallback(const WeakCallbackData<Object, TrackPromise>&);
+  inline void WeakCallback(Isolate* isolate, Local<Object> object);
+  inline TrackPromise(Isolate* isolate, Local<Object> object);
+  ~TrackPromise();
+  Persistent<Object> persistent_;
+};
```

Comment:
@bnoordhuis I'm not sure to get rid of this; it's duplicated in the header but that doesn't seem to work?